src/pymor/models/transfer_function.py

```diff
@@ -0,0 +1,473 @@
+            out = ax.semilogx(freq, 20 * np.log10(mag), **mpl_kwargs)
+        else:
+            out = ax.loglog(freq, mag, **mpl_kwargs)
+
+        ax.set_title('Magnitude plot')
+        freq_unit = ' (Hz)' if Hz else ' (rad/s)'
+        ax.set_xlabel('Frequency' + freq_unit)
+        mag_unit = ' (dB)' if dB else ''
+        ax.set_ylabel('Magnitude' + mag_unit)
+
+        return out
+
+    @cached
+    def h2_norm(self, return_norm_only=True, **quad_kwargs):
+        """Compute the H2-norm using quadrature.
+
+        This method uses `scipy.integrate.quad` and makes no assumptions on the form of the transfer
+        function.
+
+        By default, the absolute error tolerance in `scipy.integrate.quad` is set to zero (see its
+        optional argument `epsabs`).
+        It can be changed by using the `epsabs` keyword argument.
+
+        Parameters
+        ----------
+        return_norm_only
+            Whether to only return the approximate H2-norm.
+        quad_kwargs
+            Keyword arguments passed to `scipy.integrate.quad`.
+
+        Returns
+        -------
+        norm
+            Computed H2-norm.
+        norm_relerr
+            Relative error estimate (returned if `return_norm_only` is `False`).
+        info
+            Quadrature info (returned if `return_norm_only` is `False` and `full_output` is `True`).
+            See `scipy.integrate.quad` documentation for more details.
+        """
+        if not self.cont_time:
+            raise NotImplementedError
+
+        import scipy.integrate as spint
+        quad_kwargs.setdefault('epsabs', 0)
+        quad_out = spint.quad(lambda w: spla.norm(self.eval_tf(w * 1j))**2,
+                              0, np.inf,
+                              **quad_kwargs)
+        norm = np.sqrt(quad_out[0] / np.pi)
+        if return_norm_only:
+            return norm
+        norm_relerr = quad_out[1] / (2 * quad_out[0])
+        if len(quad_out) == 2:
+            return norm, norm_relerr
+        else:
+            return norm, norm_relerr, quad_out[2:]
+
+    def h2_inner(self, lti):
+        """Compute H2 inner product with an |LTIModel|.
+
+        Uses the inner product formula based on the pole-residue form
+        (see, e.g., Lemma 1 in :cite:`ABG10`).
+
+        Parameters
+        ----------
+        lti
+            |LTIModel| consisting of |Operators| that can be converted to |NumPy arrays|.
+            The D operator is ignored.
+
+        Returns
+        -------
+        inner
+            H2 inner product.
+        """
+        from pymor.models.iosys import LTIModel, _lti_to_poles_b_c
+        assert isinstance(lti, LTIModel)
+
+        poles, b, c = _lti_to_poles_b_c(lti)
+        inner = sum(c[i].dot(self.eval_tf(-poles[i]).dot(b[i]))
+                    for i in range(len(poles)))
+        inner = inner.conjugate()
+
+        return inner
+
+    def __add__(self, other):
+        assert hasattr(other, 'eval_tf') and hasattr(other, 'eval_dtf')
```

Comment:
I think it would be good to allow `eval_dtf` to be undefined for `self` and/or `other`. In such a case, the result should not have a `dtf` attribute, of course.

Reply:
True, I forgot about that.